# Tensors
A tensor is a multi-dimensional array that is the fundamental data structure used in PyTorch. We can create tensors for performing above in several ways:

In [22]:
import torch

tensor_1d = torch.tensor([1, 2, 3])
print("1D Tensor(Vector):")
print(tensor_1d)
print()

tensor_2d = torch.tensor([[1, 2], [3, 4]])
print("2D Tensor (Matrix):")
print(tensor_2d)
print()

random_tensor = torch.rand(2, 3)
print("Random Tensor (2x3):")
print(random_tensor)
print()

ones_tensor = torch.ones(2, 3)
print("Ones Tensor (2x3):")
print(ones_tensor)

1D Tensor(Vector):
tensor([1, 2, 3])

2D Tensor (Matrix):
tensor([[1, 2],
        [3, 4]])

Random Tensor (2x3):
tensor([[0.7636, 0.7486, 0.7829],
        [0.4578, 0.1423, 0.3685]])

Ones Tensor (2x3):
tensor([[1., 1., 1.],
        [1., 1., 1.]])


## Tensor Operations in PyTorch

PyTorch provides a variety of tensor operations that are essential for manipulating data efficiently, especially when preparing data for machine learning and deep learning tasks.

### Indexing

**Indexing** allows you to access specific elements within a tensor. By specifying the position of an element, you can retrieve its value directly. This is useful when working with individual data points or inspecting tensor contents.

### Slicing

**Slicing** enables you to extract a subset of a tensor by specifying a range of indices. It allows you to select particular rows, columns, or sections of a tensor, making it easier to work with specific portions of the data.

### Reshaping

**Reshaping** changes the dimensions or shape of a tensor without altering its underlying data. This operation is particularly useful when preparing data for neural networks, as different layers often require inputs in specific shapes.


In [23]:
tensor = torch.tensor([[1, 2], [3, 4], [5, 6]])

element = tensor[1, 0]
print(f"Indexed Element (Row 1, Column 0): {element}")

slice_tensor = tensor[:2, :]
print(f"Sliced Tensor (First two rows): \n{slice_tensor}")

reshaped_tensor = tensor.view(2, 3)
print(f"Reshaped Tensor (2x3): \n{reshaped_tensor}")

Indexed Element (Row 1, Column 0): 3
Sliced Tensor (First two rows): 
tensor([[1, 2],
        [3, 4]])
Reshaped Tensor (2x3): 
tensor([[1, 2, 3],
        [4, 5, 6]])


### Common Tensor Functions
PyTorch offers a variety of common tensor functions that simplify complex operations.
- Broadcasting allows for automatic expansion of dimensions to facilitate arithmetic operations on tensors of different shapes.
- Matrix multiplication enables efficient computations essential for neural network operations.

In [24]:
tensor_a = torch.tensor([[1, 2, 3], [4, 5, 6]])
tensor_b = torch.tensor([[1, 2, 3]])

boardcasted_result = tensor_a + tensor_b
print(f"Boardcasted Addition Result: \n{boardcasted_result}")

matrix_multiplication_result = torch.matmul(tensor_a, tensor_a.T)
print(f"Matrix Multiplication Result (tensor_a, tensor_b): \n{matrix_multiplication_result}")

Boardcasted Addition Result: 
tensor([[2, 4, 6],
        [5, 7, 9]])
Matrix Multiplication Result (tensor_a, tensor_b): 
tensor([[14, 32],
        [32, 77]])


## GPU Acceleration
PyTorch facilitates GPU acceleration, enabling much faster computations which is especially important in deep learning due to the extensive matrix operations involved. By transferring tensors to the GPU, you can significantly reduce training times and improve performance.

In [25]:
import torch

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {device}')

tensor_size = (10000, 10000)  
a = torch.randn(tensor_size, device=device)  
b = torch.randn(tensor_size, device=device)  

c = a + b  

print("Result shape (moved to CPU for printing):", c.cpu().shape)

print("Current GPU memory usage:")
print(f"Allocated: {torch.cuda.memory_allocated(device) / (1024 ** 2):.2f} MB")
print(f"Cached: {torch.cuda.memory_reserved(device) / (1024 ** 2):.2f} MB")

Using device: cpu
Result shape (moved to CPU for printing): torch.Size([10000, 10000])
Current GPU memory usage:
Allocated: 0.00 MB
Cached: 0.00 MB


## Building and Training Neural Networks

### Step 1: Define the Neural Network Class
In this step, we’ll define a class that inherits from torch.nn.Module. We’ll create a simple neural network with an input layer, a hidden layer and an output layer.

In [26]:
import torch
import torch.nn as nn

class SimpleNN(nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()
        self.fc1 = nn.Linear(2, 4)
        self.fc2 = nn.Linear(4, 1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x

### Step 2: Prepare the Data
Next, we’ll prepare our data. We will use a simple dataset that represents the XOR logic gate, consisting of binary input pairs and their corresponding XOR results.

In [27]:
X_train = torch.tensor([[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0]]) 
y_train = torch.tensor([[0.0], [1.0], [1.0], [0.0]])

### Step 3: Instantiate the Model, Loss Function and Optimizer
Now we will instantiate our model. We’ll also define a loss function and choose an optimizer like ***stochastic gradient descent*** to update the model’s weights based on the calculated loss.

In [28]:
import torch.optim as optim
model = SimpleNN()
criterion = nn.MSELoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

### Step 4: Training the Model
Now we enter the training loop, where we will repeatedly pass our training data through the model to learn from it.

In [29]:
for epoch in range(100):
    model.train()

    outputs = model(X_train)
    loss = criterion(outputs, y_train)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if(epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch + 1}/100], Loss: {loss.item():.4f}')

Epoch [10/100], Loss: 0.2557
Epoch [20/100], Loss: 0.2515
Epoch [30/100], Loss: 0.2480
Epoch [40/100], Loss: 0.2441
Epoch [50/100], Loss: 0.2393
Epoch [60/100], Loss: 0.2327
Epoch [70/100], Loss: 0.2237
Epoch [80/100], Loss: 0.2123
Epoch [90/100], Loss: 0.1986
Epoch [100/100], Loss: 0.1841


### Step 6: Testing the Model
Finally, we need to evaluate the model’s performance on new data to assess its generalization capability.

In [30]:
model.eval()
with torch.no_grad():
    test_data = torch.tensor([[0.0, 0.0], [0.0, 1.0], [1.0, 0.0], [1.0, 1.0]])
    predictions = model(test_data)
    print(f'Predictions:\n{predictions}')

Predictions:
tensor([[0.4005],
        [0.4450],
        [0.7141],
        [0.4248]])


## Optimizing Model Training with PyTorch Datasets
### 1. Efficient Data Handling with Datasets and DataLoaders
Dataset and DataLoader facilitates batch processing and shuffling, ensuring smooth data iteration during training.

In [31]:
import torch
from torch.utils.data import Dataset, DataLoader

class MyDataset(Dataset):
    def __init__(self):
        self.data = torch.tensor([[1.0, 2.0], [3.0, 4.0], [5.0, 6.0]])
        self.labels = torch.tensor([0, 1, 0])

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]

dataset = MyDataset()
dataloader = DataLoader(dataset, batch_size=2, shuffle=True)

for batch in dataloader:
    print("Batch Data:", batch[0])  
    print("Batch Labels:", batch[1])

Batch Data: tensor([[1., 2.],
        [5., 6.]])
Batch Labels: tensor([0, 0])
Batch Data: tensor([[3., 4.]])
Batch Labels: tensor([1])


### 2. Enhancing Data Diversity through Augmentation
Torchvision provides simple tools for applying random transformations such as rotations, flips and scaling hence enhancing the model's ability to generalize on unseen data.

In [32]:
import torchvision.transforms as transforms
from PIL import Image


image = Image.open('cat.jpg')  

transform = transforms.Compose([
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor()
])

augmented_image = transform(image)
print("Augmented Image Shape:", augmented_image.shape)

Augmented Image Shape: torch.Size([3, 180, 270])


Here is the exact breakdown of your output:
- 3 (Channels): This means your image has 3 color channels, which correspond to Red, Green, and Blue (RGB). If your image were grayscale (black and white), this number would be 1.
- 180 (Height): Your image is 180 pixels tall.
- 270 (Width): Your image is 270 pixels wide.

### 3. Batch Processing for Efficient Training
Batch processing improves computational efficiency and accelerates training, especially on hardware accelerators.

In [33]:
for epoch in range(2):  
    for inputs, labels in dataloader:
        
        outputs = inputs + 1  
        print(f"Epoch {epoch + 1}, Inputs: {inputs}, Labels: {labels}, Outputs: {outputs}")

Epoch 1, Inputs: tensor([[3., 4.],
        [5., 6.]]), Labels: tensor([1, 0]), Outputs: tensor([[4., 5.],
        [6., 7.]])
Epoch 1, Inputs: tensor([[1., 2.]]), Labels: tensor([0]), Outputs: tensor([[2., 3.]])
Epoch 2, Inputs: tensor([[3., 4.],
        [1., 2.]]), Labels: tensor([1, 0]), Outputs: tensor([[4., 5.],
        [2., 3.]])
Epoch 2, Inputs: tensor([[5., 6.]]), Labels: tensor([0]), Outputs: tensor([[6., 7.]])


## Advanced Deep Learning Models
### 1. Convolutional Neural Networks (CNNs)
- PyTorch simplifies the implementation of CNNs using modules like torch.nn.Conv2d and pooling layers.
- Integrating batch normalization with torch.nn.BatchNorm2d helps stabilize learning and accelerate training by normalizing the output of convolutional layers.
### 2. Recurrent Neural Networks (RNNs)
- Implementing RNNs in PyTorch is straightforward with torch.nn.LSTM and torch.nn.GRU modules.
- RNNs, including LSTMs and GRUs are perfect for sequential data tasks.
### 3. Generative Models
PyTorch makes it easy to construct Generative Models, including:
- Generative Adversarial Networks (GANs)
- Variational Autoencoders (VAEs)

## Transfer Learning
- Fine-Tuning Pretrained Models
- Implementing Transfer Learning with torchvision.models
- Freezing and Unfreezing Layers